core/arch/arm/pta/device.c

```diff
@@ -0,0 +1,98 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (C) 2019, Linaro Limited
+ */
+
+/*
+ * This pseudo TA is used by Linux TEE driver to fetch pseudo/early TA's UUIDs
```

Comment:
The current code will not enumerate the early TAs. I think it would be nice to have also early TAs enumerated, mybe even the installed TAs if any. In the meantime, either remove to early TA reference here or add a service in **early_ta.c** to enumare them.

Reply:
As per my previous reply, my intent was to enumerate early TAs as well but looking at complexity of fetching TA flags from elf (loading and parsing), I dropped it for now. So will remove early TA reference.

From installed TAs, are you referring to TAs with flags `TA_FLAG_INSTANCE_KEEP_ALIVE` & `TA_FLAG_SINGLE_INSTANCE` enabled? As I only see these specific TAs whose context isn't destroyed and `tee_ctxes` retain their context.

BTW, `tee_ctxes` list is easier to traverse (avoiding any duplicate pseudo TAs which are enumerated via `pseudo_tas`). If you think this is good add-on, I could add enumeration using `tee_ctxes` as well. Also it could work for loaded early TAs as well.

Reply:
Oh ok. I think enumerating user TAs (with `TA_FLAG_INSTANCE_KEEP_ALIVE`, `TA_FLAG_SINGLE_INSTANCE` and `TA_FLAG_DEVICE_ENUM` set) currently loaded in the core looks more robust solution. It could cover most of use-cases as follows:
1. Early TAs: Can be loaded in OP-TEE core during OP-TEE init as a platform specific service.
2. REE FS TAs and Secure storage TAs: Can be loaded in OP-TEE core via user-space.

Also with these flags set it makes more sense for user TA to act as Linux device (being resident and single instance). @jforissier @jenswi-linaro @etienne-lms your views?

Reply:
We know that we need to be able to enumerate pseudo TA, user TAs is still a maybe. So I think it's enough to just do the pseudo TA part now and once we know more about how user TAs will be used by the kernel we deal with those then.

Reply:
Requiring that the TA be loaded in order to be enumerated sounds a bit too restrictive to me (a service might be available but not used by the normal world, then it would be wasteful to load it always ; and you need some code to load the TA even before the "normal" user has invoked it). Same for requiring keep alive (wasteful if the TA does not need to keep a context).

I think it would be more reasonable to restrict the thing to pseudo-TAs for now.

Reply:
Should we rename `xxx_DEVICE_ENUM_xxx` into `xxx_TA_ENUM_xxx`?

Reply:
Okay will restrict this interface for pseudo-TAs only.

BTW for user TAs, if we look at this otherwise, we need to load and parse every TA elf file to get TA header flags to enumerate them. To me it also looks a bulkier and wasteful operation to perform.
Also regarding keeping alive I think it would be more robust for a Linux kernel TEE based driver rather than to depend on a user-space tee-supplicant daemon to fulfil its requests (which could be killed at run-time).